src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -51,25 +52,52 @@ public override void Initialize (AnalysisContext context)
 		{
-			if (context.ContainingSymbol.HasAttribute (RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute))
+			if (TryGetSymbolFromOperation (targetOperation, context) is not ISymbol target ||
+				TryGetSymbolFromOperation (sourceOperation, context) is not ISymbol source)
 				return;
 
-			if (TryGetSymbolFromOperation (assignmentOperation.Target) is not ISymbol target ||
-				TryGetSymbolFromOperation (assignmentOperation.Value) is not ISymbol source)
-				return;
+			CheckAndReportDynamicallyAccessedMembers (source, target, context, originOperation);
+		}
 
-			if (!target.TryGetDynamicallyAccessedMemberTypes (out var damtOnTarget))
+		static void CheckAndReportDynamicallyAccessedMembers (IOperation sourceOperation, ISymbol target, OperationAnalysisContext context, IOperation originOperation, bool targetIsMethodReturn = false)
+		{
+			if (TryGetSymbolFromOperation (sourceOperation, context) is not ISymbol source)
 				return;
 
-			source.TryGetDynamicallyAccessedMemberTypes (out var damtOnSource);
+			CheckAndReportDynamicallyAccessedMembers (source, target, context, originOperation, targetIsMethodReturn);
+		}
+
+		static void CheckAndReportDynamicallyAccessedMembers (ISymbol source, ISymbol target, OperationAnalysisContext context, IOperation originOperation, bool targetIsMethodReturn = false)
+		{
+			// For the target symbol, a method symbol may represent either a "this" parameter or a method return.
+			// The target symbol should never be a named type.
+			Debug.Assert (target.Kind is not SymbolKind.NamedType);
+			var damtOnTarget = targetIsMethodReturn
+				? ((IMethodSymbol) target).GetDynamicallyAccessedMemberTypesOnReturnType ()
+				: target.GetDynamicallyAccessedMemberTypes ();
+			// For the source symbol, a named type represents a "this" parameter and a method symbol represents a method return.
+			var damtOnSource = source.Kind switch {
+				SymbolKind.NamedType => context.ContainingSymbol.GetDynamicallyAccessedMemberTypes (),
+				SymbolKind.Method => ((IMethodSymbol) source).GetDynamicallyAccessedMemberTypesOnReturnType (),
+				_ => source.GetDynamicallyAccessedMemberTypes ()
+			};
+
 			if (Annotations.SourceHasRequiredAnnotations (damtOnSource, damtOnTarget, out var missingAnnotations))
 				return;
 
-			var diag = GetDiagnosticId (source.Kind, target.Kind);
+			var diag = GetDiagnosticId (source.Kind, target.Kind, targetIsMethodReturn);
 			var diagArgs = GetDiagnosticArguments (source.Kind == SymbolKind.NamedType ? context.ContainingSymbol : source, target, missingAnnotations);
-			context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), assignmentOperation.Syntax.GetLocation (), diagArgs));
+			context.ReportDiagnostic (Diagnostic.Create (DiagnosticDescriptors.GetDiagnosticDescriptor (diag), originOperation.Syntax.GetLocation (), diagArgs));
```

Comment:
Is `originOperation` only used for the location? If so, consider just passing along a location for the diagnostic.

Reply:
Good idea, changed to take just a location.